File: web/src/types/analytics.ts
export interface Stats {
  total_accounts: number
  total_transactions: number
  pending_transactions: number
  total_balance: number
  last_sync: string | null
}

export interface MonthlySummary {
  year: number
  month: number
  transaction_count: number
  total_amount: number
  total_charged: number
  by_status: Record<string, number>
  by_type: Record<string, number>
  by_account: Record<string, number>
}

export interface TrendPoint {
  year: number
  month: number
  total_amount: number
  transaction_count: number
}

export interface CategoryBreakdownItem {
  category: string
  count: number
  total_amount: number
  avg_amount: number
}

export interface TagBreakdownItem {
  tag: string
  count: number
  total_amount: number
  percentage: number
}

export interface CardSpendingItem {
  last4: string
  total_amount: number
  transaction_count: number
  percentage: number
}

export interface CategoryTrends {
  categories: Record<string, Array<{ year: number; month: number; amount: number }>>
  totals: Record<string, number>
}

File: api/schemas/analytics.py
"""Analytics schemas."""

from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel


class StatsResponse(BaseModel):
    total_accounts: int
    total_transactions: int
    pending_transactions: int
    total_balance: float
    last_sync: Optional[datetime] = None


class MonthlySummary(BaseModel):
    year: int
    month: int
    transaction_count: int
    total_amount: float
    total_charged: float
    by_status: Dict[str, int]
    by_type: Dict[str, int]
    by_account: Dict[str, int]


class TrendPoint(BaseModel):
    year: int
    month: int
    total_amount: float
    transaction_count: int


class CategoryBreakdownItem(BaseModel):
    category: str
    count: int
    total_amount: float
    avg_amount: float


class TagBreakdownItem(BaseModel):
    tag: str
    count: int
    total_amount: float
    percentage: float


class CardSpendingItem(BaseModel):
    last4: str
    total_amount: float
    transaction_count: int
    percentage: float


class CategoryTrendsResponse(BaseModel):
    categories: Dict[str, List[Dict[str, Any]]]
    totals: Dict[str, float]
